### ISSUE TITLE
No way to create exclusive fullscreen window using LWJGL3 backend

### ISSUE THREAD

**Issue by Schyfis** [libgdx/libgdx#4785]
#### Issue details
Using the LWJGL2 backend, the default behavior for setting a window to fullscreen was to make the window "exclusive fullscreen." When in exclusive fullscreen mode, a window catches the mouse and typically gives better performance.
In the LWJGL3 backend, the default behavior has changed. Making a window fullscreen now causes it to go into "borderless fullscreen." In borderless fullscreen mode, the mouse is not caught and the game typically performs worse than exclusive fullscreen. To my knowledge, there is no way to set an exclusive fullscreen display mode through the libGDX API- any attempt to set a fullscreen mode will only result in a borderless fullscreen window.

Why does this matter? In my case I'm able to run my game smoothly, but recording gameplay footage in fullscreen slows down the game significantly, while the LWJGL2 backend had no issues. Based on this, I imagine that users with worse hardware might have issues running the game smoothly in a borderless fullscreen window.

#### Reproduction steps/code
Simply make any game using the LWJGL3 backend fullscreen, using either the Lwjgl3ApplicationConfiguration or `Gdx.graphics.setFullscreenMode(mode)`. If you have multiple monitors, you can tell if the window is exclusive fullscreen based on whether or not it catches the mouse by default.

#### Version of LibGDX and/or relevant dependencies
1.9.6, LWJGL3 backend. This issue has been present since the introduction of the LWJGL3 backend.

#### Please select the affected platforms
- [ ] Android
- [ ] iOS (robovm)
- [ ] iOS (MOE)
- [ ] HTML/GWT
- [X] Windows
- [ ] Linux
- [ ] MacOS

I'm unsure if the same results are produced on Linux and Mac.

**Comment by Tom-Ski** [comment]
@code-disaster got any tips on this?

**Comment by CreamyCookie** [comment]
I think this could be useful, but it should be optional. I'm not sure, for example, if I like the mouse to get caught.

**Comment by Tom-Ski** [comment]
Yup, fullscreen borderless and fullscreen exclusive would be great to have.

**Comment by Schyfis** [comment]
@CreamyCookie I totally agree, they should both be available. In the LWJGL2 backend they were both available, but it was a little wonky. For borderless fullscreen you had to do something like this before you start the app: `System.setProperty("org.lwjgl.opengl.Window.undecorated", "true");`

It'd be nice to have something like `Gdx.graphics.setFullscreenMode(DisplayMode mode, boolean exclusive)` for the LWJGL3 backend. It'd be cleaner and provide access to borderless and exclusive modes.

**Comment by code-disaster** [comment]
See http://www.glfw.org/docs/latest/window_guide.html#window_creation, the *"Windowed full screen" windows* section

GLFW only uses this "windowed full screen" mode if you request a video mode which matches the desktop mode (size & refresh rate). For all other resolutions, a full screen exclusive mode is used.

The LWJGL3 backend happens to [ignore the refresh rate](https://github.com/libgdx/libgdx/blob/master/backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/Lwjgl3Application.java#L438), for a reason probably even @badlogic who's done this doesn't remember.

When I uncomment this line, on my PC (NVidia 1070), for any refresh rate != desktop setting, it switches to exclusive (the monitor flickers on change), but it seems to ignore the number I give and always defaults to 60 hz.

**Note** that in any case, and any resolution, the mouse cursor isn't caught. You may need to track the mouse position and use Input.setCursorPosition(). I'm not aware of any GLFW function which keeps the system cursor visible and captured to the window area.

**Comment by FredGithub** [comment]
I agree the mode achieved by calling `setFullscreenMode` must be some kind of weird hybrid. I can't believe GLFW thought it was a good idea to make it non exclusive if size and refresh rate match the monitor's, to allow for "faster window creation and application switching". For games, the focus is performance, not "fast application switching"...

However, borderless fullscreen is something else: it's windowed mode, except you hide the borders of the window and make it the size of the monitor. You can alt-tab to another window, and it will show in front of your game, which will still be visible in the background. 

If you call `setFullscreenMode`, and alt-tab to another window, notice that your game will completely hide, confirming that it was indeed in some kind of weird fullscreen mode.

Window mode has to be the most confusing and buggy part of game programming by far, with each library and OS doing what it wants and using different terms. And yet players expect us to support all resolutions, all refresh rates, any count of monitors, on all OSes perfectly 😄

**Comment by paulrayrussell** [comment]
Is this still an issue pls? I'm investigating frameworks and want to create a full screen  native desktop app at various resolutions including the is current one.

**Comment by Schyfis** [comment]
It's been a while but I updated my LWJGL3 project and can confirm that **YES, it is still not possible to create an exclusive fullscreen window using the LWJGL3 backend.**

However, regarding one of the issues in my opening post:

> Why does this matter? In my case I'm able to run my game smoothly, but recording gameplay footage in fullscreen slows down the game significantly, while the LWJGL2 backend had no issues. Based on this, I imagine that users with worse hardware might have issues running the game smoothly in a borderless fullscreen window.

This seems to be resolved! I can record footage of my updated LWJGL3 project without any slowdowns. One more step towards being usable!

**Comment by xGnoSiSx** [comment]
I need to comment here, because I have the same experience on my project. There is no way to set exclusive fullscreen video mode for the LWJGL3 backend. I didn't wish to open a seperate issue.

On my dual monitor setup, I'm having even more worrisome issues with it:
GTX 980m Optimus with Intel integrated GPU, dual monitor setup (laptop + external)

1. Setting *any* mode on my primary monitor with libgdx, changes resolution on the monitor and I always get a borderless windowed mode.

2. Pressing alt-tab doesn't bring other windows to the front.

3. Pressing alt-tab doesn't do anything to suspend/resume the application - I can't test my resume code that recreates the context - app never looses the context.

4. Shift + windows key + left or right keys DO move the window to the other monitor as expected for a windowed mode.

My only recourse, unless I find a work around, is to fall back to LWJGL2.0 (undesirable but unavoidable otherwise)

**Comment by tommyettinger** [comment]
I have alt-tab working well enough (on a single monitor) just by enabling fullscreen and auto-iconify in the launcher.

```java
        Lwjgl3ApplicationConfiguration configuration = new Lwjgl3ApplicationConfiguration();
        configuration.setResizable(true); // not sure if needed
        configuration.useVsync(true);
        configuration.setFullscreenMode(Lwjgl3ApplicationConfiguration.getDisplayMode());
        configuration.setAutoIconify(true);
```

`setAutoIconify(true)` seems to be needed because without it, the window just returns to the foreground as soon as the alt-tab is selected. I don't have any multi-monitor machine, so I can't really test how this works there. The problem with this approach is that with some screen capture software, like ShareX, only the window behind the game is captured while the game is in fullscreen.

There's a possibly-viable alternative, also only using configuration in Lwjgl3's launcher:
```java
        configuration.setResizable(true); // almost certainly needed!
        configuration.useVsync(true); // not sure if needed.
        configuration.setWindowListener(new Lwjgl3WindowAdapter() {
            @Override
            public void maximized(boolean isMaximized) {
                if(isMaximized)
                    Gdx.graphics.setFullscreenMode(Gdx.graphics.getDisplayMode());
                else
                    Gdx.graphics.setWindowedMode(YOUR_GAME_WIDTH_CONSTANT, YOUR_GAME_HEIGHT_CONSTANT);
            }

            @Override
            public void focusLost() {
                Gdx.app.postRunnable(() -> maximized(false));
            } on
        });
```
This is weird; it starts the app as windowed (in a window YOUR_GAME_WIDTH_CONSTANT wide by YOUR_GAME_HEIGHT_CONSTANT high), and if the user interacts with the OS to maximize, it fullscreens only then. For whatever reason, the user interaction makes ShareX correctly capture the fullscreen game and only that game; Alt-Tab works as well in this case. I think this has to do with the display parameters being different during a maximize operation than during initial launch configuration, but that's just a guess.

**Comment by code-disaster** [comment]
side note: I see more consistent and stable results across desktop platforms if I _always_ launch the game in windowed mode, then switch to the target / fullscreen mode according to user settings at load time.

**Comment by noblemaster** [comment]
I think `autoIconify` should be enabled by default on the backend. Without it, ALT+TAB doesn't work. I don't see any downsides to having it enabled by default? Or am I missing something? Having it disabled (as of now) just causes problems with ALT+TAB.

What exclusive fullscreen mode concerns: Does it have to be exclusive? The current fullscreen mode appears to be working just fine? Did anyone actually check if performance is that much worse between exclusive vs. non-exclusive? Is LWJGL(v.2)'s exclusive mode rendering much faster than LWJGL3's non-exclusive mode? What are other benefits of exclusive fullscreen mode? 

I actually prefer the non-exclusive full-screen mode. It's less irritating as the OS isn't trying to actually change the screen resolution of the screen and all the other windows are still fine. It's faster also switching from/to fullscreen mode.

**Comment by crykn** [comment]
> @noblemaster: I think `autoIconify` should be enabled by default on the backend. [...] I don't see any downsides to having it enabled by default?

I completely agree. 

> @noblemaster: What exclusive fullscreen mode concerns: Does it have to be exclusive? [...] Did anyone actually check if performance is that much worse between exclusive vs. non-exclusive? Is LWJGL(v.2)'s exclusive mode rendering much faster than LWJGL3's non-exclusive mode?

I'm wondering about that as well. If anyone has a proper game and could try that out under real-world conditions, it would be much appreciated. This could also shed some light on the notion that LWJGL3 is _in general_ less performant than LWJGL2, as is suspected by some users on Discord.

> @noblemaster: What are other benefits of exclusive fullscreen mode?

No idea. 

Catching the mouse pointer is something that should be done manually in my opinion. Locking the cursor to the center of the window is possible via `Gdx.input.setCursorCatched(true);`, and confining the cursor to the window constraints is being worked on, see https://github.com/glfw/glfw/issues/58. For the time being, the latter can also be achieved by hiding the cursor and drawing one yourself (this is recommended for example [here](https://discourse.glfw.org/t/locking-cursor-to-window-without-hiding-it/869/2) and [here](https://discuss.cocos2d-x.org/t/confine-mouse-cursor-to-window/34377/2)).

Moreover, when `autoIconify` is enabled, the current fullscreen mode behaves like I would expect an excluisve fullscreen mode to (and very different to a borderless window fullscreen): other windows can't be on top and the game is minimized when I click on another monitor. 

In addition, even if we wanted, I don't think there is anything we can do to enable exclusive fullscreen. This comment (https://github.com/glfw/glfw/issues/1036#issuecomment-311498678) seems to suggest that there is no difference at the API level. In line with this, I wasn't able to confirm the findings of @code-disaster that using a different refresh rate [here](https://github.com/libgdx/libgdx/blob/03489663718ccba1f25a55c9fd829b3f4eecf745/backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/Lwjgl3Application.java#L438) changed anything, but tbh I am also not really sure how I would be able to distinguish the two.

**To sum it up: I'm not sure whether there are any advantages to using exclusive fullscreen, but regardless of that, I don't think we can enable it anyway. Apart from that, we should enable `autoIconify` by default.**

**Comment by crykn** [comment]
Since there seem to be no objections, I'm going to close this issue. 

For future readers, the following steps should work:

- If you want **borderless window fullscreen**, use this:
   ```java
   Graphics.DisplayMode displayMode = Lwjgl3ApplicationConfiguration.getDisplayMode();
   config.setWindowedMode(displayMode.width, displayMode.height);
   config.setDecorated(false);
   ```
- If you want (GLFW's somewhat) **exclusive fullscreen** (i.e. other windows can't be on top and the application is minimized when clicking on another monitor), use this:
   ```java
   config.setAutoIconify(true);
   config.setFullscreenMode(Lwjgl3ApplicationConfiguration.getDisplayMode());
   ```
   If you run into any issues with screen capturing not working, try out @tommyettinger's solution [above](https://github.com/libgdx/libgdx/issues/4785#issuecomment-739088752).

**Comment by Daburnell112** [comment]
This needs to be fixed... I'm getting complaints from my testers that they can't actually use "real fullscreen mode"(™) after switching to Lwjgl3. There's no way to do this at runtime either with that solution, because while the game is running, I don't have a way as far as I can see to call "Gdx.graphics.setAutoIconify(true)" with an existing Lwjgl3Application window if I start it in borderless. If there's a workaround, let me know, but I've done a ton of research, and I can't find a proper solution for this.

**Comment by Frosty-J** [comment]
> There's no way to do this at runtime either

I'll preface this by saying I dislike the `autoIconify` behaviour.

* If you want full screen where Windows Alt+Tab works, set the window as undecorated and to the monitor's screen resolution (can maximise it or set size directly).
  * Caution: Order matters if doing this after the game has been created! Behaviour will differ between Windows and Linux if you get the order wrong.
* If you want PrintScreen to work properly and no black flash when the OS draws something else (such as volume indicator) on Windows Intel drivers, increase the window size by 1 on either axis (warning: can spill over into other monitors with multi-monitor setups).
  * Especially as this can spill over into other monitors for multi-monitor users, you may wish to only extend the screen when necessary with `if (UiUtils.isWindows && Gdx.gl.glGetString(GL20.GL_RENDERER).contains("Intel"))`.
* If you want actual exclusive full screen, you'll probably have to stick with LWJGL2 (though I just tried it and it crashed my graphics, requiring disabling and re-enabling the graphics card at the OS level to fix, so your mileage may vary).

**With that out the way**, you should be able to do something like this, updating the boolean how you please. I've excluded the other overrides for brevity:

```java
final boolean[] myRealFullscreenFlag = { true };

config.setWindowListener(new Lwjgl3WindowListener() {
  Lwjgl3Window window;

  @Override
  public void created(Lwjgl3Window window) {
    this.window = window;
  }

  @Override
  public void focusLost() {
    if (Gdx.graphics.isFullscreen() && myRealFullscreenFlag[0])
      window.iconifyWindow();
  }
}
```

**Comment by Daburnell112** [comment]
> If you want actual exclusive full screen, you'll probably have to stick with LWJGL2 (though I just tried it and it crashed my graphics, requiring disabling and re-enabling the graphics card at the OS level to fix, so your mileage may vary).

Was working fine for me in an earlier version, had it running in my game for about 5 years with LWJGL 2. One of the things I'm considering is somehow using LWJGL 2 in Fullscreen when switching to that mode - but the newer LWJGL 3 doubles my frame rate on my project so really I can't go back to the old version.

That being said the behavior with auto-iconify doesn't make it go into real fullscreen mode™ at all, at least with the build on Maven. At the very least I can use borderless, disable the Fullscreen option in my project, but the mouse still exits the screen on a multi monitor setup and I'm having to use the AWT Robot class to keep the mouse within the bounds of the window, something I'm still working on. If I figure that out, I'll post the code here for posterity.

**Comment by Frosty-J** [comment]
> the mouse still exits the screen on a multi monitor setup

Can also be solved in the window listener or with your game's `Lwjgl3Graphics#getWindow`.

```java
glfwSetInputMode(window.getWindowHandle(), GLFW_CURSOR, GLFW_CURSOR_CAPTURED);
```

**Comment by jeffcragg** [comment]
Hi anybody interested in this, full screen exclusive mode is dropped from Win10 onwards which is why GLFW doesn't support. For more info see https://devblogs.microsoft.com/directx/demystifying-full-screen-optimizations/
